File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int kthSmallest(TreeNode root, int k) 
    {
        PriorityQueue<Integer> pq = new PriorityQueue<>((a,b) -> b-a);
        kthSmallestHelper(root,k,pq);
        return pq.peek();
        
    }
    public void kthSmallestHelper(TreeNode root, int k, PriorityQueue<Integer> pq){
        if(root == null) return;
        pq.add(root.val);
        if(pq.size() > k){
            pq.poll();
        }
        
        kthSmallestHelper(root.left,k, pq);
        kthSmallestHelper(root.right,k, pq);

    }
}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def helper(self, preorder, inorder, start, end):
        if end < start:
            return None
        i = 0
        for j in range(start, end+1):
            if inorder[j] == preorder[self.index]:
                i = j
                break
        
        root = TreeNode(preorder[self.index],None,None)
        self.index += 1
        root.left = self.helper(preorder, inorder, start, i-1)
        root.right = self.helper(preorder, inorder, i+1, end)
        return root


    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        self.index = 0
        return self.helper(preorder, inorder, 0, len(preorder)-1)

        